WikiWatershed/model-my-watershed
# Split MapShed GIS and GWLF-E Endpoints

## Overview

We need to run the MapShed GIS chain only once per project, and make it available to all the scenarios in the project. Each scenario will run the GWLF-E model with the result of the original MapShed GIS chain run, plus any modifications.

Thus, we split the current chain in two: one to run the MapShed GIS gathering step (`start_mapshed`),  and another to run the GWLF-E model on it (`start_gwlfe`).

Initially we had thought of saving the MapShed GIS Dictionary to the database. However since we need to support transient anonymous users whose data is not saved in the database, we return the entire dictionary to the browser instead, and expect it to be returned in entirety to the GWLF-E endpoint. If and when this data-transfer becomes too big and problematic, we will research ideas for saving it.

Since numpy arrays are not easily serialized and deserialization will be a matter of guessing whether the array should be a numpy array or not, we switch to using plain lists instead. Since this was the only place where numpy was used, we remove the dependency entirely. GWLF-E may still use numpy internally.

Optionally, if we discover that numpy is essential for computational performance of the model, we also record a list of each array field in the DataModel. If needed, we can convert all Python lists to be numpy arrays after deserialization and before running the model as simply as:

```python
import numpy as np
from django.conf import settings
from gwlfe import gwlfe, DataModel

# Inside some task which actually runs the model
data = json.loads(input_json)
for field in settings.GWLFE_CONFIG['ArrayFields']:
    data[field] = np.array(data[field])

z = DataModel(data)
output = gwlfe.run(z)

return output
```

~~This PR should not be merged until #1274, #1290, and #1277 have been merged and this work updated to account for them.~~ Those issues have been merged, thus this PR is now ready for review.

## Testing Instructions

Run `./scripts/debugcelery.sh`. Create a GWLF-E job in the MMW UI. You should see a GWLF-E run in Celery and the output of the saved sample file in the UI.

To test the `mapshed` endpoint, we must use the command line since there isn't a UI for it yet (that's coming in #1310). Thus, with [httpie](http://httpie.org) installed, submit a job to trigger MapShed GIS gathering:

```http
$ http --form :8000/api/modeling/start/mapshed/ mapshed_input='{"area_of_interest":{"type":"MultiPolygon","coordinates":[[[[-75.18805468442467,39.95683776295151],[-75.17632519838781,39.95683776295151],[-75.17632519838781,39.96582815672415],[-75.18805468442467,39.96582815672415],[-75.18805468442467,39.95683776295151]]]]}}'
HTTP/1.1 200 OK
Allow: POST, OPTIONS
Connection: keep-alive
Content-Encoding: gzip
Content-Type: application/json
Date: Wed, 18 May 2016 03:46:27 GMT
Server: nginx/1.10.0
Transfer-Encoding: chunked
Vary: Accept-Encoding
Vary: Accept, Cookie

{
    "job": "cb8c279a-8250-4de9-8e91-6a2e5f5a9895", 
    "status": "started"
}
```

Once the Celery run is complete, you should be able to query for the job:

```http
$ http :8000/api/modeling/jobs/cb8c279a-8250-4de9-8e91-6a2e5f5a9895/
HTTP/1.1 200 OK
Allow: OPTIONS, GET
Connection: keep-alive
Content-Encoding: gzip
Content-Type: application/json
Date: Wed, 18 May 2016 03:47:16 GMT
Server: nginx/1.10.0
Transfer-Encoding: chunked
Vary: Accept-Encoding
Vary: Accept, Cookie

{
    "error": "", 
    "finished": "2016-05-18T03:46:30.433Z", 
    "job_uuid": "cb8c279a-8250-4de9-8e91-6a2e5f5a9895", 
    "result": "......", 
    "started": "2016-05-18T03:46:25.651Z", 
    "status": "complete"
}
```

Connects #1296
